File: src/java/spring/src/main/java/com/benchmark/controller/JsonController.java
package com.benchmark.controller;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Map;

@RestController
public class JsonController {
    @GetMapping("/json")
    public Map<String, Object> json() {
        var items = new ArrayList<Map<String, Object>>();
        String timestamp = Instant.now().toString();

        for (int i = 0; i < 1000; i++) {
            items.add(Map.of(
                "id", i,
                "name", "User " + i,
                "email", "user" + i + "@example.com",
                "timestamp", timestamp
            ));
        }

        return Map.of("items", items, "count", items.size(), "timestamp", timestamp);
    }
}

File: src/kotlin/spring/src/main/kotlin/com/benchmark/controller/JsonController.kt
package com.benchmark.controller

import com.benchmark.Canonical

import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
class JsonController {

    /**
     * The previous implementation emitted {id,name,email,active,tags} -- a
     * shape no other implementation used, with a three-element list per item
     * that inflated the payload -- and ignored ?n=.
     */
    @GetMapping("/json")
    fun json(@RequestParam(name = "n", required = false) n: String?): Map<String, Any> =
        Canonical.response(n)
}
